doc/decisions/1_in_discussion/internal_cache.md

```diff
@@ -0,0 +1,350 @@
+# Internal KeySet Cache
+
+## Problem
+
+`kdbGet` might return more or fewer keys than requested.
+This was found confusing several times.
+
+### More Keys
+
+Even if calling `kdbGet` with a parent key below a mountpoint, `kdbGet` will nevertheless return all keys of the mountpoint.
+Pseudo code example, assuming there is a mountpoint at `/mountpoint` and a key `/mountpoint/other`:
+
+```
+kdbGet (kdb, ks, keyNew("/mountpoint/below"));
+assert (ksLookup (ks, "/mountpoint/other") == NULL);
+```
+
+It was found unexpected that this assert will fail.
+
+### Fewer Keys
+
+When doing a second `kdbGet` with a new keyset no keys will be returned when no backends report changed data, because kdb internally thinks the data is already up-to-date.
+A unit test by @atmaxinger:
+
+```c
+static void test_doubleGet (void)
+{
+	printf("running %s\n", __func__);
+
+	// Setup
+	Key * parentKey = keyNew("/somewhere", KS_END);
+	KeySet * ks = ksNew(0, KS_END);
+	KDB * kdb = kdbOpen (ksNew(0, KS_END), parentKey);
+	kdbGet (kdb, ks, parentKey);
+	ksAppendKey (ks, keyNew ("user:/somewhere", KEY_VALUE, "abc", KEY_END));
+	ksAppendKey (ks, keyNew ("user:/somewhere/key", KEY_VALUE, "xyz", KEY_END));
+	kdbSet (kdb, ks, parentKey);
+	kdbClose (kdb, parentKey);
+
+	// Scenario
+	kdb = kdbOpen (ksNew(0, KS_END), parentKey);
+
+	KeySet * ks1 = ksNew (0, KS_END);
+	KeySet * ks2 = ksNew (0, KS_END);
+
+	kdbGet (kdb, ks1, keyNew("/somewhere", KEY_END));
+	succeed_if (ksLookupByName (ks1, "/somewhere/key", 0) != NULL, "should find key (1)");
+	kdbGet (kdb, ks2, keyNew("/somewhere", KEY_END));
+	succeed_if (ksLookupByName (ks2, "/somewhere/key", 0) != NULL, "should find key (2)");
+}
+```
+
+> It actually outputs should find key (2) so the assertion fails.
```

Comment:
This is also said in the next sentence.

Reply:
[0734b46](https://github.com/ElektraInitiative/libelektra/pull/4641/commits/0734b46b19d14ec65636f3428c65435d6f2bb06f)